Q2. What is the difference between Bernoulli Naive Bayes and Multinomial Naive Bayes?

Solution :

The difference between Bernoulli Naive Bayes and Multinomial Naive Bayes is that Multinomial Naive Bayes works with occurrence counts while Bernoulli works with binary/boolean features.

Multinomial Naive Bayes is widely used for document classification and keeps the count of frequent words present in the documents. Bernoulli Naive Bayes is used for discrete data, where features are only in binary form.

Q3. How does Bernoulli Naive Bayes handle missing values?

Solution :

In Naive Bayes, omitting records with any missing values or omitting only the missing attributes are two ways to handle missing values. When calculating probabilities, the best way to handle missing values is to ignore that category if it is a missing value. Essentially, you are saying: only compute a probability based on the specific categories I have provided with non-missing values.

Q4. Can Gaussian Naive Bayes be used for multi-class classification?

Solution :

Yes, Gaussian Naive Bayes can be used for multi-class classification. It is a variant of Naive Bayes that is used for continuous data. It assumes that the features are normally distributed and calculates the mean and variance of each feature for each class. It then uses these values to calculate the probability of each class given the input features.

Q5. Assignment: Spambase dataset

SOlution :

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

import warnings
warnings.filterwarnings('ignore')

In [7]:
df  = pd.read_csv('spambase.data' , header=None)

In [8]:
df

0     1     2    3     4     5     6     7     8     9   ...     48  \
0     0.00  0.64  0.64  0.0  0.32  0.00  0.00  0.00  0.00  0.00  ...  0.000   
1     0.21  0.28  0.50  0.0  0.14  0.28  0.21  0.07  0.00  0.94  ...  0.000   
2     0.06  0.00  0.71  0.0  1.23  0.19  0.19  0.12  0.64  0.25  ...  0.010   
3     0.00  0.00  0.00  0.0  0.63  0.00  0.31  0.63  0.31  0.63  ...  0.000   
4     0.00  0.00  0.00  0.0  0.63  0.00  0.31  0.63  0.31  0.63  ...  0.000   
...    ...   ...   ...  ...   ...   ...   ...   ...   ...   ...  ...    ...   
4596  0.31  0.00  0.62  0.0  0.00  0.31  0.00  0.00  0.00  0.00  ...  0.000   
4597  0.00  0.00  0.00  0.0  0.00  0.00  0.00  0.00  0.00  0.00  ...  0.000   
4598  0.30  0.00  0.30  0.0  0.00  0.00  0.00  0.00  0.00  0.00  ...  0.102   
4599  0.96  0.00  0.00  0.0  0.32  0.00  0.00  0.00  0.00  0.00  ...  0.000   
4600  0.00  0.00  0.65  0.0  0.00  0.00  0.00  0.00  0.00  0.00  ...  0.000   

         49   50     51     52     53     54   55    56  57  
0     0.000  0.0  0.778  0.000  0.000  3.756   61   278   1  
1     0.132  0.0  0.372  0.180  0.048  5.114  101  1028   1  
2     0.143  0.0  0.276  0.184  0.010  9.821  485  2259   1  
3     0.137  0.0  0.137  0.000  0.000  3.537   40   191   1  
4     0.135  0.0  0.135  0.000  0.000  3.537   40   191   1  
...     ...  ...    ...    ...    ...    ...  ...   ...  ..  
4596  0.232  0.0  0.000  0.000  0.000  1.142    3    88   0  
4597  0.000  0.0  0.353  0.000  0.000  1.555    4    14   0  
4598  0.718  0.0  0.000  0.000  0.000  1.404    6   118   0  
4599  0.057  0.0  0.000  0.000  0.000  1.147    5    78   0  
4600  0.000  0.0  0.125  0.000  0.000  1.250    5    40   0  

[4601 rows x 58 columns]

In [9]:
df = df.rename(columns = {57 : 'target'})

In [10]:
df.head()

0     1     2    3     4     5     6     7     8     9  ...    48  \
0  0.00  0.64  0.64  0.0  0.32  0.00  0.00  0.00  0.00  0.00  ...  0.00   
1  0.21  0.28  0.50  0.0  0.14  0.28  0.21  0.07  0.00  0.94  ...  0.00   
2  0.06  0.00  0.71  0.0  1.23  0.19  0.19  0.12  0.64  0.25  ...  0.01   
3  0.00  0.00  0.00  0.0  0.63  0.00  0.31  0.63  0.31  0.63  ...  0.00   
4  0.00  0.00  0.00  0.0  0.63  0.00  0.31  0.63  0.31  0.63  ...  0.00   

      49   50     51     52     53     54   55    56  target  
0  0.000  0.0  0.778  0.000  0.000  3.756   61   278       1  
1  0.132  0.0  0.372  0.180  0.048  5.114  101  1028       1  
2  0.143  0.0  0.276  0.184  0.010  9.821  485  2259       1  
3  0.137  0.0  0.137  0.000  0.000  3.537   40   191       1  
4  0.135  0.0  0.135  0.000  0.000  3.537   40   191       1  

[5 rows x 58 columns]

In [11]:
# dependent and independent features
X = df.drop('target' , axis=1)
y = df['target']

In [12]:
# train test split
from sklearn.model_selection import train_test_split

In [13]:
X_train,X_test,y_train,y_test = train_test_split(X,y , test_size=0.30 , random_state=0)

In [14]:
X_train.shape , X_test.shape

((3220, 57), (1381, 57))

In [90]:
from sklearn.naive_bayes import GaussianNB

In [91]:
gnb = GaussianNB()

In [92]:
gnb.fit(X_train , y_train)

GaussianNB()

In [93]:
y_pred = gnb.predict(X_test)

In [94]:
from sklearn.metrics import confusion_matrix , accuracy_score

In [95]:
matrix = confusion_matrix(y_test ,y_pred)

In [96]:
matrix

array([[597, 225],
       [ 34, 525]])

In [97]:
accuracy = accuracy_score(y_test,y_pred)

In [98]:
accuracy

0.8124547429398986

In [99]:
t_pos = matrix[0][0]
f_pos = matrix[0][1]
f_neg = matrix[1][0]
t_neg = matrix[1][1]

In [100]:
precision = t_pos / (t_pos + f_pos)

In [101]:
precision  # important

0.7262773722627737

# K Fold cross validation [Gaussian naive bayes]

In [85]:
from sklearn.model_selection import KFold
from sklearn.naive_bayes import GaussianNB

In [86]:
gnb = GaussianNB()

In [83]:
cv = KFold(n_splits=10 , shuffle=True)

In [84]:
from sklearn.model_selection import cross_val_score

In [87]:
score_gnb = cross_val_score(gnb , X_train,y_train,scoring='accuracy',cv=cv)

In [88]:
score_gnb

array([0.82919255, 0.82298137, 0.84161491, 0.82608696, 0.83229814,
       0.82919255, 0.81055901, 0.83229814, 0.82919255, 0.85093168])

In [89]:
np.mean(score_gnb)

0.8304347826086957

# Multinomial Naive bayes

# K fold

In [31]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.model_selection import KFold

In [34]:
mnb = MultinomialNB()

In [35]:
k = KFold(n_splits=10 , shuffle=True)

In [33]:
from sklearn.model_selection import cross_val_score

In [37]:
scores = cross_val_score(mnb ,X_train, y_train , scoring='accuracy', cv=k)

In [38]:
scores

array([0.81677019, 0.80124224, 0.80745342, 0.83850932, 0.80745342,
       0.80434783, 0.80434783, 0.77950311, 0.80124224, 0.7484472 ])

In [39]:
np.mean(scores)

0.8009316770186334

In [58]:
mnb = MultinomialNB()

In [59]:
mnb.fit(X_train,y_train)

MultinomialNB()

In [60]:
y_pred_mnb = mnb.predict(X_test)

In [70]:
mat_mnb = confusion_matrix(y_test,y_pred_mnb)

In [71]:
mat_mnb

array([[699, 123],
       [140, 419]])

In [72]:
accuracy_mnb = accuracy_score(y_test,y_pred_mnb)

In [73]:
accuracy_mnb

0.8095582910934106

In [74]:
t_pos_mnb = mat_mnb[0][0]
f_pos_mnb = mat_mnb[0][1]
f_neg_mnb = mat_mnb[1][0]
t_neg_mnb = mat_mnb[1][1]

In [80]:
precision_mnb = t_pos_mnb / (t_pos_mnb + f_pos_mnb)

In [81]:
precision_mnb

0.8503649635036497

# Bernoulii naive bayes

In [40]:
from sklearn.naive_bayes import BernoulliNB

In [41]:
bnb = BernoulliNB()

In [42]:
cv = KFold(n_splits=10 , shuffle=True)

In [43]:
cv

KFold(n_splits=10, random_state=None, shuffle=True)

In [44]:
score = cross_val_score(bnb,X_train,y_train,scoring='accuracy',cv=cv)

In [46]:
score

array([0.90993789, 0.85714286, 0.90372671, 0.88198758, 0.87888199,
       0.88819876, 0.90993789, 0.9068323 , 0.84782609, 0.88509317])

In [47]:
np.mean(score)

0.8869565217391304

In [48]:
bnb = BernoulliNB()

In [49]:
bnb.fit(X_train,y_train)

BernoulliNB()

In [50]:
y_pred_bnb = bnb.predict(X_test)

In [63]:
mat_bnb = confusion_matrix(y_test,y_pred_bnb)

In [64]:
mat_bnb

array([[769,  53],
       [119, 440]])

In [65]:
accuracy_bnb = accuracy_score(y_test,y_pred_bnb)

In [66]:
accuracy_bnb

0.8754525706010138

In [67]:
t_pos_bnb = mat_bnb[0][0]
f_pos_bnb = mat_bnb[0][1]
f_neg_bnb = mat_bnb[1][0]
t_neg_bnb = mat_bnb[1][1]

In [78]:
precision_bnb = t_pos_bnb / (t_pos_bnb + f_pos_bnb)

In [79]:
precision_bnb

0.9355231143552312

## conclusion

 # Bernoulii naive bayes gives the best accuracy of 87.5 % 